automata/fa/dfa.py

```diff
@@ -529,6 +548,93 @@ def isfinite(self):
         except nx.exception.NetworkXNoCycle:
             return True
 
+    def count_words_of_length(self, k):
+        """
+        Counts words of length k assuming state is the initial state
+        """
+        self._ensure_count_for_length(k)
+        return self._count_cache[k][self.initial_state]
+
+    def _ensure_count_for_length(self, k):
+        """
+        Populate count cache up to length k
+        """
+        while len(self._count_cache) <= k:
+            i = len(self._count_cache)
+            self._count_cache.append(defaultdict(int))
+            if i == 0:
+                for state in self.final_states:
+                    self._count_cache[i][state] = 1
+            else:
+                for state in self.states:
+                    for _, suffix_state in self.transitions[state].items():
+                        self._count_cache[i][state] += self._count_cache[i-1][suffix_state]
+ 
+    def words_of_length(self, k):
+        """
+        Generates all words of size k in the language represented by the DFA
+        """
+        self._ensure_words_of_length(k)
+        for word in self._word_cache[k][self.initial_state]:
+            yield word
+
+    def _ensure_words_of_length(self, k):
+        """
+        Populate word cache up to length k
+        """
+        while len(self._word_cache) <= k:
+            i = len(self._word_cache)
+            self._word_cache.append(defaultdict(set))
+            if i == 0:
+                for state in self.final_states:
+                    self._word_cache[i][state].add('')
+            else:
+                for state in self.states:
+                    for symbol, suffix_state in self.transitions[state].items():
+                        for word in self._word_cache[i-1][suffix_state]:
+                            self._word_cache[i][state].add(symbol + word)
```

Comment:
This entry in the cache is accessed a lot, and `i` is not modified in this loop. You should factor out some of these lookups, as they can get expensive.

Reply:
Good point, updated